### PR TITLE
Add Pallas SparseCore guide

### PR DESCRIPTION
**Pull request by IvyZX**

### CODE REVIEW

File: docs/pallas/tpu/sparsecore.md [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+
+A walkthrough on each of the components:
+
+* **SC Tiles (SCT)**: The vector processing subcores of a SparseCore. Each subcore has its own memory, so the data flow is independent.
```
Comment (superbobry) [inline docs/pallas/tpu/sparsecore.md]: Should we use the Pallas SC terminology here: SC vector subcores (and SC scalar subcores) below?

File: docs/pallas/tpu/sparsecore.md [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+
+  * In Pallas, the mesh of SCTs are denoted as `plsc.VectorSubcoreMesh`.
+
+* **Lanes (SIMD width)**: A SC tile do computation on a size-N vector in "Single instruction multiple data (SIMD)" fashion. That means it runs the same computation on all the N numbers in the input lane at the same time.
+
+* **SC Sequencer (SCS)**: The scalar processing subcore of a SparseCore. It is capable of scalar operations, dynamic indexing as well as instruction fetching & dispatch to tiles.
```
Comment (superbobry) [inline docs/pallas/tpu/sparsecore.md]: Is it worth mentioning that the scalar subcore can also do DMAs?

File: docs/pallas/tpu/sparsecore.md [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+
+  * In Pallas, the mesh SCS is denoted as `plsc.ScalarSubcoreMesh`.
+
+* **Memory spaces**: Each SC tile has its own VMEM and SMEM (omitted in the graph) space. They also have access to a shared VMEM space. Similar to Tensorcore, these VMEM spaces connect with the TPU's HBM.
+
+  * In Pallas, the VMEM spaces are denoted as `pltpu.VMEM` and `pltpu.VMEM_SHARED`, and SMEM is denoted as `pltpu.SMEM`.
+
+  * In some other documentations, the shared VMEM could be called "Spmem", and the per-subcore VMEM called "TileSpmem" or "local Spmem".
```
Comment (superbobry) [inline docs/pallas/tpu/sparsecore.md]: I think it's best to leave this out here. Let's focus on the terminology used in Pallas SC/Mosaic SC.

File: docs/pallas/tpu/sparsecore.md [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+
+  * SparseCore subcores can also start data transfers to and from the VMEMs in TensorCore.
```
Comment (superbobry) [inline docs/pallas/tpu/sparsecore.md]: This is only  true for some TPU generations I think. Should we mention that?

File: docs/pallas/tpu/sparsecore.md [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+print(sc_info)
+```
+
+    SparseCore info for TPU 7x:
+    SparseCoreInfo(num_cores=2, num_subcores=16, num_lanes=16, dma_granule_size_bytes=64)
+
+
+## Operations and workloads
+
+A SparseCore consists of 16 smaller processing units, each with its own data flow. That makes it good for workloads that have the following characteristics:
```
Comment (superbobry) [inline docs/pallas/tpu/sparsecore.md:68]: We have already explained the processing units above, so I'd say we should just call them "vector subcores" or "SC tiles" (if you want to keep this terminology). E.g.

"Each vector subcore on the SparseCore has its own dataflow. That makes ..."

File: docs/pallas/tpu/sparsecore.md [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+
+* Medium-to-low amount of computation
+
+* Frequent data communications
+
+Some of the useful operations on SparseCore are:
+
+* Small vector arithmetics
+
+* Indexed fetch & send (gather and scatter)
```
Comment (superbobry) [inline docs/pallas/tpu/sparsecore.md]: Wdyt about saying gather/scatter right away instead of "Indexed fetch & send"?

File: docs/pallas/tpu/sparsecore.md [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+
+* Sorting, unique, counts, histograms
+
+* Ragged operations
+
+## Express SparseCore hardware
+
+Similar to in TensorCore, Pallas uses mesh to express the compute units in SparseCore. You should create a `ScalarSubcoreMesh` for SCS, and a `VectorSubcoreMesh` for SC tiles.
```
Comment (superbobry) [inline docs/pallas/tpu/sparsecore.md]: As above, I have a mild preference towards using "scalar/vector subcore" instead of SCS/SC tiles.

File: docs/pallas/tpu/sparsecore.md [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+```
+
+    135 µs ± 10.8 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
+    105 µs ± 8.8 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
+    200 µs ± 647 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
+
+
+## Gather and scatter
+
+SparseCore has specific optimized ops for indexed retrievals and overwrites. Given an input or output ref in HBM (named `data`) and an array of indices in VMEM (named `indices`), it can quickly read from ("gather") or write to ("scatter") `data[indices]`.
```
Comment (superbobry) [inline docs/pallas/tpu/sparsecore.md]: Nit: "and updates"? since we also have `add=True` in `pltpu.async_copy`.

File: docs/pallas/tpu/sparsecore.md [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+
+A walkthrough on each of the components:
+
+* **SC Tiles (SCT)**: The vector processing subcores of a SparseCore. Each subcore has its own memory, so the data flow is independent.
+
+  * In Pallas, the mesh of SCTs are denoted as `plsc.VectorSubcoreMesh`.
+
+* **Lanes (SIMD width)**: A SC tile do computation on a size-N vector in "Single instruction multiple data (SIMD)" fashion. That means it runs the same computation on all the N numbers in the input lane at the same time.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: Nit: "An SC vector subcore does computation on size-N vectors in a "Single instruction multiple data (SIMD) fashion. Computations will happen on all the numbers in a lane in a single instruction.

File: docs/pallas/tpu/sparsecore.md (hunk 2 above)
Comment (sharadmv) [reply docs/pallas/tpu/sparsecore.md]: +1

File: docs/pallas/tpu/sparsecore.md (hunk 4 above)
Comment (sharadmv) [reply docs/pallas/tpu/sparsecore.md]: I think it's not worth mentioning atm (since we don't support it in Pallas yet).

File: docs/pallas/tpu/sparsecore.md [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+* **Memory spaces**: Each SC tile has its own VMEM and SMEM (omitted in the graph) space. They also have access to a shared VMEM space. Similar to Tensorcore, these VMEM spaces connect with the TPU's HBM.
+
+  * In Pallas, the VMEM spaces are denoted as `pltpu.VMEM` and `pltpu.VMEM_SHARED`, and SMEM is denoted as `pltpu.SMEM`.
+
+  * In some other documentations, the shared VMEM could be called "Spmem", and the per-subcore VMEM called "TileSpmem" or "local Spmem".
+
+  * SparseCore subcores can also start data transfers to and from the VMEMs in TensorCore.
+
+
+Actual specs vary by TPU versions. Some of the published TPU specs are listed below:
```
Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: versions -> version

Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: Here are some published TPU specs:

File: docs/pallas/tpu/sparsecore.md [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+print(tiles_mesh)
+```
+
+    ScalarSubcoreMesh(axis_name='core', num_cores=2)
+    VectorSubcoreMesh(core_axis_name='core', subcore_axis_name='subcore', num_cores=2, num_subcores=16)
+
+
+## A basic SparseCore kernel
+
+See below for a simple SCS kernel that includes DMAs, per-core customizing and compute operations. Note that SCS kernels can only do scalar operations.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: I'd indicate that here, we are mapping the program over two scalar subcores (because there are two sparsecores). I'd also initialize `x` to have a leading dimension of `num_cores`, not 2.

File: docs/pallas/tpu/sparsecore.md [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+
+  return kernel(x)
+
+x = jnp.arange(sc_info.num_lanes * 2).reshape((2, sc_info.num_lanes))
+np.testing.assert_array_equal(cumsum(x), jnp.cumsum(x, axis=1))
+```
+
+## Pipelining in SparseCore kernels
+
+You can `pltpu.emit_pipeline` to write pipelined SparseCore kernels. This parallelizes your code on all subcores of all (or some) SparseCores.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: This parallelizes your code on all subcores of all (or some) SparseCores. -> The `core_axis_name` and `dimension_semantics` arguments to `emit_pipeline` enable partitioning the pipeline across sparsecores/subcores.

File: docs/pallas/tpu/sparsecore.md [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+
+You can label your data axis as `pltpu.PARALLEL` for `emit_pipeline` to automatically parallelize work along it. See an example below. You can also stick with `axis_index` to customize per-core behavior (example [here](https://docs.jax.dev/en/latest/pallas/tpu/core_map.html#mapping-over-sparsecores)).
```
Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: I'd delete this paragraph, and say after the example code:

Alternatively, you can use `axis_index` to compute the core index and use it to split up work across cores (example [here]...).

File: docs/pallas/tpu/sparsecore.md [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+    135 µs ± 10.8 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
+    105 µs ± 8.8 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
+    200 µs ± 647 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
+
+
+## Gather and scatter
+
+SparseCore has specific optimized ops for indexed retrievals and overwrites. Given an input or output ref in HBM (named `data`) and an array of indices in VMEM (named `indices`), it can quickly read from ("gather") or write to ("scatter") `data[indices]`.
+
+Inside a Pallas kernel, this can be done by initializing `sync_copy` or `async_copy` directly with `data_ref[indices_ref]`.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: This should say `data_ref.at[indices_ref]` not `data_ref[indices_ref]`.

Also, can we reword to:

We can use these gather/scatter by indexing a Ref with an indices Ref as part of an `async_copy` or `sync_copy`. For example,  `sync_copy(data_ref.at[indices_ref], target_ref)` will trigger a gather.

File: docs/pallas/tpu/sparsecore.md [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+
+Below shows a pipelined kernel that does indexed retrieval by putting a gather operation at the body.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: Nit: Below is a kernel that pipelines loading indices into a vector subcore's VMEM and in the body, we execute a gather using those indices.

File: docs/pallas/tpu/sparsecore.md [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+
+gathered = jnp.take(data, indices, axis=0)
+out = scatter(gathered, indices)
+np.testing.assert_array_equal(out, data)
+
+```
+
+## Benchmark against TensorCore
+
+Gather and scatter are operations within SparseCore's strengths. We can implement the same in naive JAX APIs, which default to run on TensorCore, and compare the results.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: Sparsecores are particularly good at gather and scatter operations. We can implement the same using vanilla JAX APIs, which by default run on Tensorcore and compare the results.

File: docs/pallas/tpu/sparsecore.md (hunk 1 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done, renamed the image & phrases to subcores. I kept a mention of "tile" since it's a common term in existing XLA docs.

File: docs/pallas/tpu/sparsecore.md (hunk 2 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

File: docs/pallas/tpu/sparsecore.md (hunk 3 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: The multiple names among Pallas/Mosaic/XLA lands are confusing for anybody new (it certainly confused me), so I want this Internet to have at least one paragraph that clarifies and connects them all :)

File: docs/pallas/tpu/sparsecore.md (hunk 4 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

File: docs/pallas/tpu/sparsecore.md (hunk 5 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md:68]: Done.

File: docs/pallas/tpu/sparsecore.md (hunk 6 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

File: docs/pallas/tpu/sparsecore.md (hunk 8 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

File: docs/pallas/tpu/sparsecore.md (hunk 9 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

File: docs/pallas/tpu/sparsecore.md (hunk 10 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

File: docs/pallas/tpu/sparsecore.md (hunk 11 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

File: docs/pallas/tpu/sparsecore.md (hunk 12 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

File: docs/pallas/tpu/sparsecore.md (hunk 13 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

File: docs/pallas/tpu/sparsecore.md (hunk 14 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

File: docs/pallas/tpu/sparsecore.md (hunk 15 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

File: docs/pallas/tpu/sparsecore.md (hunk 16 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: Done.

File: docs/pallas/tpu/sparsecore.md [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+from jax.experimental.pallas import tpu_sc as plsc
+import jax.numpy as jnp
+import numpy as np
+
+assert pltpu.get_tpu_info().sparse_core is not None, "No SparseCore found"
+```
+
+## Hardware overview
+
+Depending on the version, a recent TPU chip may have 2 or 4 SparseCores. A SparseCore consists of multiple subcores, each having its own VMEM space. Below is a diagram for a SparseCore inside the TPU.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: "a recent TPU chip" -> "a TPU chip"

File: docs/pallas/tpu/sparsecore.md [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+
+A walkthrough on each of the components:
+
+* **Vector subcore (tiles)**: The vector processing subcores of a SparseCore. Each subcore has its own memory, so the data flow is independent.
+
+* **Lanes (SIMD width)**: An SC vector subcore does computation on size-N vectors in a "Single instruction multiple data (SIMD) fashion. Computations will happen on all the numbers in a lane in a single instruction.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: Missing endquote on Single instruction multiple data. Also nit: let's Single -> single

File: docs/pallas/tpu/sparsecore.md [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+
+* **Scalar subcore**: The scalar processing subcore of a SparseCore. It is capable of scalar operations, dynamic indexing as well as initiating DMAs and streams.
+
+* **Memory spaces**: Each vector subcore has its own VMEM and SMEM (omitted in the graph) space. They also have access to a shared VMEM space. The scalar subcore has its own SMEM. All these spaces connect with the TPU's HBM.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: graph -> graphic

File: docs/pallas/tpu/sparsecore.md (hunk 3 above)
Comment (sharadmv) [reply docs/pallas/tpu/sparsecore.md]: Nit on wording:

In other documentation, the shared VMEM is often called "SPMEM" and the per-subcore VMEM is called "TileSPMEM" or "local SPMEM".

I do think in the long run, we will want different names across SC and TC memory spaces, so we'll update this when we settle on it.

File: docs/pallas/tpu/sparsecore.md [hunk 20]
```diff
@@ -0,0 +1,345 @@
...
+        def _(c1):
+          slc = (pl.ds(c0, SC_REG_OP_SHAPE[0]), pl.ds(c1, SC_REG_OP_SHAPE[1]))
+          out_vmem.at[*slc][...] = in_vmem.at[*slc][...] + 1
+
+    pltpu.emit_pipeline(
+        sc_add_one_body,
+        grid=(in_shape[0] // dma_block[0], in_shape[1] // dma_block[1]),
+        in_specs=[pl.BlockSpec(block_shape=dma_block, index_map=lambda i, j: (i, j))],
+        out_specs=[pl.BlockSpec(block_shape=dma_block, index_map=lambda i, j: (i, j))],
+        core_axis_name='subcore',
```
Comment (sharadmv) [inline docs/pallas/tpu/sparsecore.md]: Can we use just one sparsecore in this kernel? Otherwise we are not parallelizing across sparsecores here.

**Review by sharadmv — APPROVED** [review]

File: docs/pallas/tpu/sparsecore.md (hunk 20 above)
Comment (IvyZX) [reply docs/pallas/tpu/sparsecore.md]: I set `core_axis_name=('core', 'subcore')` and it still worked (same run time). Does that mean this already parallelizes across all sparsecores?
